docs/keto/modeling/learn-the-concepts.mdx

```diff
@@ -0,0 +1,295 @@
+
+class Document implements Namespace {
+  // highlight-start
+  related: {
+    owners: User[]
+    editors: User[]
+    viewers: User[]
+    parents: Folder[]
+  }
+  // highlight-end
+}
+
+class Folder implements Namespace {
+  // highlight-start
+  related: {
+    owners: User[]
+    editors: User[]
+    viewers: User[]
+    parents: Folder[]
+  }
+  // highlight-end
+}
+```
+
+### Define permissions in the OPL
+
+The permissions are expressed in the OPL as TypeScript functions that take a context containing the subject, for example the
+current user, and that answer the permission check based on the relations this object has, in this case the current document.
+
+:::tip
+
+The permissions from the description are declared as functions inside the `permits` property of the corresponding class in the Ory
+Permission Language.
+
+:::
+
+Let's see this in action for the first permission:
+
+- `view` a document if the user is a `viewer`, `editor`, or `owner` of the document; or if the user can `view` the parent folder
+
+```ts title="permissions-v3.ts"
+import { Namespace, Context } from "@ory/keto-namespace-types"
+
+class User implements Namespace {}
+
+class Document implements Namespace {
+  related: {
+    owners: User[]
+    editors: User[]
+    viewers: User[]
+    parents: Folder[]
+  }
+
+  permits = {
+    //highlight-start
+    view: (ctx: Context): boolean =>
+      this.related.viewers.includes(ctx.subject) ||
+      this.related.editors.includes(ctx.subject) ||
+      this.related.owners.includes(ctx.subject) ||
+      this.related.parents.traverse((parent) => parent.permits.view(ctx)),
+    //highlight-end
+  }
+}
+
+class Folder implements Namespace {
+  related: {
+    owners: User[]
+    editors: User[]
+    viewers: User[]
+    parents: Folder[]
+  }
+}
+```
+
+Let's do this for the remaining permissions:
+
+```ts title="permissions-v4.ts"
+import { Namespace, Context } from "@ory/keto-namespace-types"
+
+class User implements Namespace {}
+
+class Document implements Namespace {
+  related: {
+    owners: User[]
+    editors: User[]
+    viewers: User[]
+    parents: Folder[]
+  }
+
+  permits = {
+    view: (ctx: Context): boolean =>
+      this.related.viewers.includes(ctx.subject) ||
+      this.related.editors.includes(ctx.subject) ||
+      this.related.owners.includes(ctx.subject) ||
+      this.related.parents.traverse((parent) => parent.permits.view(ctx)),
+    //highlight-start
+    edit: (ctx: Context): boolean =>
+      this.related.editors.includes(ctx.subject) ||
+      this.related.owners.includes(ctx.subject) ||
+      this.related.parents.traverse((parent) => parent.permits.edit(ctx)),
+    delete: (ctx: Context): boolean =>
+      this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.delete(ctx)),
+    share: (ctx: Context): boolean =>
+      this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.share(ctx)),
```

Comment:
In this case we have kind of hierarchical permissions, i.e. `delete = share > edit > view`. IMO it would be best practice to implement this hierarchy also in the permissions, as that ensures the hierarchy is always preserved. Therefore, permission escalations due to misconfiugration are less likely.

This however is only the case for hierarchical permissions, so it depends on the use-case if that is valid. Should we do it here or just keep as is?

Reply:
Hmm, I was also thinking about a section to refactor the permission model, showing the benefit of it being "mostly TypeScript". Then you would write

```ts
  permits = {
￼    view: (ctx: Context): boolean =>
￼      this.related.viewers.includes(ctx.subject) ||
￼      this.permits.edit(ctx),
￼  // ...
```
and so on.

Reply:
Yes exactly. This has the advantage that you automatically include one permission in the other, which makes it less likely that there is an error. But that only holds for cases like this where you can actually do that.